File: pre_vertafore/mitch/queue.py
class Node:
    def __init__(self, value, pointer):
        self.pointer = pointer
        self.value = value

class Queue:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0
    
    def enqueue(self, value):
        newNode = Node(value, None)
        self.size += 1
        if self.tail:
            self.tail.pointer = newNode
        self.tail = newNode
        
        if not self.head:
            self.head = newNode
            return self
    
    def dequeue(self):
        if self.head:
            self.size -= 1
            temp = self.head
            self.head = self.head.pointer
            return temp.value
        else:
            print "Queue is empty, can't dequeue"
            
    def getSize(self):
        return int(self.size)
            
            
q = Queue()
q.enqueue(1)
q.enqueue(2)
q.enqueue(3)
print q.getSize()
print q.dequeue()
print q.getSize()

nameslist = ["Abra", "Billy", "Cadabra", 'Dinky', "Effe"]
def hotPotato(names, num):
    myQueue = Queue()
    for name in names:
        myQueue.enqueue(name)

    while myQueue.getSize() > 1:
        for i in range(num):
            print i
            myQueue.enqueue(myQueue.dequeue())
        myQueue.dequeue()

    return myQueue.dequeue()

print hotPotato(nameslist, 6)

File: ctci/3. StacksAndQueues/queue.js
const Node = require('./node').Node;

class Queue{
    constructor(){
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    enqueue(value){
        let newNode = new Node(value);
        this.size++;
        if (this.head == null){
            this.head = newNode;
            this.tail = this.head;
            return this;
        } else {
            this.tail.next = newNode;
            this.tail = newNode;
            return this;
        }
    }

    dequeue(){
        if (this.head){
            let ret = this.head;
            this.head = this.head.next;
            if (this.tail == ret){
                this.tail = null;
            }
            this.size--;
            return ret;
        } else {
            console.log("Can't dequeue, queue is empty!");
            return this;
        }
    }

    getSize(){
        return this.size;
    }

    peek(){
        if (this.head){
            return this.head.value;
        } else {
            console.log("Empty queue, can't peek!");
        }
    }

    print(){
        let strList = "["
        let current = this.head
        while (current) {
            strList += current.getValue();
            if (current.getNext()){
                strList += ", "
            }
            current = current.getNext();
        }
        strList += "]"
        console.log(strList);
    }
}

// var myqueue = new Queue();
// myqueue.enqueue(1);
// myqueue.dequeue();
// myqueue.dequeue();
// myqueue.peek();
// myqueue.enqueue(2);
// myqueue.enqueue(3);
// myqueue.peek();
// myqueue.print();
// console.log(myqueue.getSize());

module.exports = {
    Queue: Queue
}